hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java

```diff
@@ -53,18 +53,18 @@ public ReadManyOperation(String name, long startSequence, int minSize, int maxSi
 
     @Override
     public void beforeRun() {
-        RingbufferContainer ringbuffer = getRingBufferContainer();
-        ringbuffer.checkBlockableReadSequence(startSequence);
     }
 
     @Override
     public boolean shouldWait() {
         if (resultSet == null) {
-            resultSet = new ReadResultSetImpl<O, O>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
+            resultSet = new ReadResultSetImpl<>(minSize, maxSize, getNodeEngine().getSerializationService(), filter);
             sequence = startSequence;
         }
 
         RingbufferContainer ringbuffer = getRingBufferContainer();
+        sequence = ringbuffer.clampReadSequenceToBounds(sequence);
+
         if (minSize == 0) {
             if (!ringbuffer.shouldWait(sequence)) {
```

Comment:
@mmedenjak commented the opposite at some point, but yes, I think this is more appropriate, except as `sequence < ringbuffer.tailSequence() + 1` (_smaller than_, instead of _not equal_). Will modify.

Reply:
Between `ringbuffer.shouldWait(sequence)` and `sequence != ringbuffer.tailSequence() + 1`, I don't mind which one we use. So I'm ok with the above change. But, let's try to analyse if `ringbuffer.shouldWait(sequence)` can, in fact, throw an exception.

In case the ringbuffer store is not configured, `ringbuffer.clampReadSequenceToBounds(sequence)` will change the `sequence` to a valid sequence in the range `[headSequence, tailSequence + 1]`. In the case when the ringbuffer store is enabled, it will not fix any sequences below `headSequence` because it assumes that the store has older sequences.

After clamping, we invoke `ringbuffer.shouldWait(sequence)`. The only method here which may throw an exception is `RingbufferContainer#checkBlockableReadSequence` and in two cases:
• `isTooLargeSequence(readSequence)` returns true -> this means `sequence > tailSequence() + 1` but this is not possible because we cover this case in `clampReadSequenceToBounds` and there is no concurrent mutation on the ringbuffer since it's all done in the partition thread
• `isStaleSequence(readSequence)` returns true -> this means `sequence < headSequence() && !store.isEnabled()` returns true. This too is not possible because it too is covered by the clamping: either the sequence is equal or larger than the head sequence or the store is enabled - such is the contract of the `clampReadSequenceToBounds` method. 

In summary, I think `ringbuffer.shouldWait(sequence)` can throw an exception after `ringbuffer.clampReadSequenceToBounds(sequence)` is invoked but I don't mind using 
`sequence < ringbuffer.tailSequence() + 1` instead.